### ISSUE TITLE
terrible elixir performance

### ISSUE THREAD

**Issue by pankamyk** [nvim-treesitter/nvim-treesitter#1957]
### Before reporting

- [X] I have updated my neovim version to latest _master_
- [X] I have updated my plugin to the latest version
- [X] I have run `:TSUpdate`

### Describe the bug

Neovim take anywhere from 750ms up to a second to open any elixir file, be it a simple test file or a large module.

### To Reproduce

1. Install elixir treesitter parser
2. Open any file
3. Notice the large lag

### Expected behavior

Obviously some additional delay is expected, but this much makes elixir with treesitter basically unusable

### Output of `:checkhealth nvim-treesitter`

```markdown
## Installation
  - OK: `tree-sitter` found  0.20.0 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v14.18.1 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (GCC) 11.1.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - comment        ✓ . . . . 
  - javascript     ✓ ✓ ✓ ✓ ✓ 
  - json           ✓ ✓ ✓ ✓ . 
  - bash           ✓ ✓ ✓ . ✓ 
  - html           ✓ ✓ ✓ ✓ ✓ 
  - css            ✓ . ✓ ✓ ✓ 
  - elixir         ✓ ✓ ✓ ✓ ✓ 
  - heex           ✓ . ✓ ✓ ✓ 
  - typescript     ✓ ✓ ✓ ✓ ✓ 
  - c              ✓ ✓ ✓ ✓ ✓ 
  - tsx            ✓ ✓ ✓ ✓ ✓ 
  - vue            ✓ . ✓ . ✓ 
  - rust           ✓ ✓ ✓ ✓ ✓ 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.6.0-dev+525-ge921e98ce
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-11 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovim/neovim/build/include
Compiled by runner@fv-az50-602

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/share/nvim"

Run :checkhealth for more info
```

### Additional context

Here's the startup log when opening an elixir file:

> times in msec
>  clock   self+sourced   self:  sourced script
>  clock   elapsed:              other lines
> 
> 000.008  000.008: --- NVIM STARTING ---
> 000.261  000.253: locale set
> 000.567  000.306: inits 1
> 000.580  000.013: window checked
> 000.619  000.039: parsing arguments
> 000.656  000.037: expanding arguments
> 000.693  000.037: inits 2
> 001.072  000.379: init highlight
> 001.074  000.002: waiting for UI
> 001.746  000.673: done waiting for UI
> 001.755  000.008: init screen for UI
> 001.892  000.015  000.015: sourcing /usr/share/nvim/archlinux.vim
> 001.897  000.046  000.031: sourcing /etc/xdg//nvim/sysinit.vim
> 004.009  002.054  002.054: sourcing /home/pankamyk/.local/share/nvim/site/autoload/plug.vim
> 012.010  000.027  000.027: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/elixir.vim
> 012.044  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/fennel.vim
> 012.076  000.016  000.016: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/fish.vim
> 012.104  000.016  000.016: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/gdresource.vim
> 012.130  000.012  000.012: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/gdscript.vim
> 012.153  000.009  000.009: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/gomod.vim
> 012.184  000.016  000.016: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/graphql.vim
> 012.215  000.017  000.017: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/hcl.vim
> 012.240  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/heex.vim
> 012.264  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/hjson.vim
> 012.287  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/json5.vim
> 012.312  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/julia.vim
> 012.340  000.016  000.016: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/ledger.vim
> 012.365  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/nix.vim
> 012.389  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/ql.vim
> 012.428  000.023  000.023: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/query.vim
> 012.460  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/surface.vim
> 012.485  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/teal.vim
> 012.508  000.009  000.009: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/tlaplus.vim
> 012.532  000.010  000.010: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/yang.vim
> 012.557  000.011  000.011: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/ftdetect/zig.vim
> 012.678  000.012  000.012: sourcing /usr/share/vim/vimfiles/ftdetect/PKGBUILD.vim
> 013.064  006.651  006.358: sourcing /usr/share/nvim/runtime/filetype.vim
> 013.279  000.037  000.037: sourcing /usr/share/nvim/runtime/ftplugin.vim
> 013.488  000.030  000.030: sourcing /usr/share/nvim/runtime/indent.vim
> 014.154  000.303  000.303: sourcing /usr/share/nvim/runtime/syntax/syncolor.vim
> 014.373  000.602  000.299: sourcing /usr/share/nvim/runtime/syntax/synload.vim
> 014.454  000.819  000.217: sourcing /usr/share/nvim/runtime/syntax/syntax.vim
> 021.070  000.135  000.135: sourcing /usr/share/nvim/runtime/syntax/syncolor.vim
> 021.853  000.151  000.151: sourcing /usr/share/nvim/runtime/syntax/syncolor.vim
> 022.394  000.128  000.128: sourcing /usr/share/nvim/runtime/syntax/syncolor.vim
> 023.654  002.855  002.442: sourcing /home/pankamyk/.config/nvim/plugged/shady.vim/colors/shady.vim
> 024.055  022.137  009.691: sourcing /home/pankamyk/.config/nvim/init.vim
> 024.065  000.127: sourcing vimrc file(s)
> 024.514  000.223  000.223: sourcing /home/pankamyk/.config/nvim/plugged/vim-transparent/plugin/background.vim
> 025.019  000.430  000.430: sourcing /home/pankamyk/.config/nvim/plugged/vim-eunuch/plugin/eunuch.vim
> 025.449  000.185  000.185: sourcing /home/pankamyk/.config/nvim/plugged/vim-gitgutter/autoload/gitgutter/utility.vim
> 025.822  000.186  000.186: sourcing /home/pankamyk/.config/nvim/plugged/vim-gitgutter/autoload/gitgutter/highlight.vim
> 026.734  001.613  001.242: sourcing /home/pankamyk/.config/nvim/plugged/vim-gitgutter/plugin/gitgutter.vim
> 026.841  000.029  000.029: sourcing /home/pankamyk/.config/nvim/plugged/nvim-toggle-terminal/plugin/nvim_toggle_terminal.vim
> 027.767  000.867  000.867: sourcing /home/pankamyk/.config/nvim/plugged/fzf.vim/plugin/fzf.vim
> 029.700  001.854  001.854: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/plugin/nvim-treesitter.vim
> 030.006  000.235  000.235: sourcing /home/pankamyk/.config/nvim/plugged/playground/plugin/nvim-treesitter-playground.vim
> 030.528  000.260  000.260: sourcing /usr/share/nvim/runtime/plugin/gzip.vim
> 030.558  000.012  000.012: sourcing /usr/share/nvim/runtime/plugin/health.vim
> 030.651  000.080  000.080: sourcing /usr/share/nvim/runtime/plugin/man.vim
> 030.680  000.014  000.014: sourcing /usr/share/nvim/runtime/plugin/matchit.vim
> 030.831  000.139  000.139: sourcing /usr/share/nvim/runtime/plugin/matchparen.vim
> 030.866  000.019  000.019: sourcing /usr/share/nvim/runtime/plugin/netrwPlugin.vim
> 031.020  000.141  000.141: sourcing /usr/share/nvim/runtime/plugin/rplugin.vim
> 031.136  000.099  000.099: sourcing /usr/share/nvim/runtime/plugin/shada.vim
> 031.182  000.024  000.024: sourcing /usr/share/nvim/runtime/plugin/spellfile.vim
> 031.338  000.141  000.141: sourcing /usr/share/nvim/runtime/plugin/tarPlugin.vim
> 031.439  000.083  000.083: sourcing /usr/share/nvim/runtime/plugin/tohtml.vim
> 031.474  000.019  000.019: sourcing /usr/share/nvim/runtime/plugin/tutor.vim
> 031.667  000.177  000.177: sourcing /usr/share/nvim/runtime/plugin/zipPlugin.vim
> 032.320  000.567  000.567: sourcing /usr/share/vim/vimfiles/plugin/fzf.vim
> 032.753  001.661: loading plugins
> 032.846  000.093: loading packages
> 033.027  000.182: loading after plugins
> 033.038  000.011: inits 3
> 037.712  004.673: reading ShaDa
> 041.252  000.019  000.019: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/after/ftplugin/elixir.vim
> 088.514  000.013  000.013: sourcing /home/pankamyk/.config/nvim/plugged/nvim-treesitter/after/ftplugin/elixir.vim
> 092.971  055.228: opening buffers
> 093.159  000.130  000.130: sourcing /home/pankamyk/.config/nvim/plugged/vim-gitgutter/autoload/gitgutter.vim
> 093.422  000.321: BufEnter autocommands
> 093.425  000.004: editing files in windows
> 093.524  000.056  000.056: sourcing /home/pankamyk/.config/nvim/plugged/vim-transparent/autoload/background.vim
> 095.779  002.298: VimEnter autocommands
> 095.783  000.004: UIEnter autocommands
> 096.092  000.228  000.228: sourcing /usr/share/nvim/runtime/autoload/provider/clipboard.vim
> 096.100  000.089: before starting main loop
> 754.147  658.047: first screen update
> 754.154  000.006: --- NVIM STARTED ---

**Comment by jeroenbourgois** [comment]
I have exactly the same experience, it is horrible. Is there something I can do to help investigate, debug, ...? Be glad to help!

**Comment by theHamsta** [comment]
@jonatanklosko

**Comment by jonatanklosko** [comment]
Hey! To clarify, parsing itself is definitely not a problem. I tested against the elixir repository and parsing a single file takes around 2ms on average. So the problem lies in querying the parsed AST (could be highlighting, but maybe some other feature that uses queries), which may be more expensive than usually due to the nested nature of the AST.

The question is if there's a specific bit that has huge impact on the performance, or if it's a tiny difference that accumulates. It would make sense to try commenting out all [queries](https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/elixir) and then gradually add them (starting from highlighting), to try figure out what could be the bottleneck. @connorlay can probably give more insights on how to go about doing that :)

**Comment by theHamsta** [comment]
Sure, it's definitely not a parsing issue but rather querying or some flaw in our highlighter,  maybe I should have tagged @connorlay in the first place

**Comment by connorlay** [comment]
> The question is if there's a specific bit that has huge impact on the performance, or if it's a tiny difference that accumulates

Based on my experience working with the previous Elixir parser and the current one, I think it is more of an accumulation of highlighting queries that slows things down, which makes me think this is a performance issue with the highlighter itself that manifests with Elixir due to the nested nature of the AST. JavaScript, by comparison, has a very flat AST.

Given that the Elixir highlights we currently have closely follow those in the main Elixir grammar repository, I recommend folks supply their own custom highlight queries to achieve the desired performance until the highlighter itself can be investigated.

Alternatively, the highlighting module can be disabled for Elixir and the regex-based highlights can be used from [vim-elixir](https://github.com/elixir-editors/vim-elixir).

**Comment by alemuller** [comment]
Tree-sitter parsing is fast, but querying is still slow. This can be seem running a few command, see https://github.com/tree-sitter/tree-sitter/discussions/1467#discussioncomment-1565138.

If you want to investigate, the query is done at `tree-sitter/lib/src/query.c`.

**Comment by connorlay** [comment]
Update on this issue, as I'm experiencing it myself now: I believe this is an issue in neovim itself and is caused by the recent update to tree-sitter 0.20.1: https://github.com/neovim/neovim/commit/e1cea4b38da4a39678ebcb0f0b638f0c13df082d

Building neovim from the previous commit https://github.com/neovim/neovim/commit/64abd7be7993ed3fa031b5d71a4482871108093a solves the regression for me locally.

**Comment by theHamsta** [comment]
Please report this to upstream tree-sitter when it can be reproduced from tree-sitter CLI

**Comment by connorlay** [comment]
@jonatanklosko I was able to reproduce this in `tree-sitter-elixir` by upgrading `tree-sitter-cli` from `0.19` to `0.20` (matching neovim) and noticed a significant (~6x) slowdown when running the syntax highlighter tests.

@theHamsta Do you think this is at all related to https://github.com/tree-sitter/tree-sitter/issues/1469?

**Comment by connorlay** [comment]
After a bit of digging, I think I found where the bottleneck is: https://github.com/tree-sitter/tree-sitter/blob/master/lib/src/query.c#L772 which changed recently with this commit https://github.com/tree-sitter/tree-sitter/commit/26dac9b2dd52c9c3998f600cad85dc7d4b63974a

Here is a visual stacktrace produced via using `perf` and `flamegraph` with `nvim foo.ex +qall` (inspired by https://github.com/neovim/neovim/issues/15461)
![image](https://user-images.githubusercontent.com/6548350/143670670-b6f68adc-edb0-4f8f-9aaa-de3a013a0549.png)

**Comment by theHamsta** [comment]
@connorlay can you report your insight at tree-sitter?

**Comment by theHamsta** [comment]
@connorlay I actually thought that we ported the query caching mechanism to upstream nvim. But apparently not for the highlighter https://github.com/theHamsta/neovim/blob/4f1d5b9751ecf7dda5fe70132dd01c8337d0da60/runtime/lua/vim/treesitter/highlighter.lua#L99-L97 / https://github.com/theHamsta/neovim/blob/4f1d5b9751ecf7dda5fe70132dd01c8337d0da60/runtime/lua/vim/treesitter/query.lua

**Comment by connorlay** [comment]
@theHamsta It looks like queries explicitly set via `M.set_query` are [cached after parsing](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/treesitter/query.lua#L129) in `M.get_query`. Are you suggesting something similar for `"highlights"` specifically, or all queries in general? Additionally, could the parsed queries be cached to disk so they persist between neovim sessions?

**Comment by theHamsta** [comment]
@clason would it be worth to continue using 0.20.0 for the 0.6 release while we're figuring out what's going on? The version bump was pretty close to the actual release date.

**Comment by clason** [comment]
Hmmm... Tree-sitter is still experimental, so this is something we can address in 0.6.1, too. I'd rather not downgrade to a pinned commit again -- especially since distros will (and already have) ignored that; homebrew for example has been building against 0.20.1 from day one.

**Comment by bigardone** [comment]
@kiuKisas same problem here.
What I have done in the meantime is:
- Uninstall TS elixir.
- Add `elixir` to `ignore_install` in the setup options.
- Installed https://github.com/elixir-editors/vim-elixir for syntax highlighting, formatting, etc...
I hope it helps :)

**Comment by alexander-p30** [comment]
@kiuKisas @bigardone 

Another workaround is compiling neovim with ```tree-sitter``` ```0.20.0``` by changing ```TREESITTER_URL``` and ```TREESITTER_C_URL``` in ```CMakeLists.txt```, this way you still get the ```tree-sitter``` highlight.

It's been working good for me so far :)

**Comment by clason** [comment]
Can you give the linked PR a try?

**Comment by rverton** [comment]
It's a lot faster than before the fix for me here, however there is still a noticeable delay.

```


times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.009  000.009: --- NVIM STARTING ---
000.822  000.812: locale set
008.884  008.062: inits 1
008.906  000.023: window checked
018.533  009.627: parsing arguments
018.817  000.284: expanding arguments
018.919  000.102: inits 2
019.710  000.791: init highlight
019.717  000.007: waiting for UI
023.577  003.859: done waiting for UI
023.591  000.014: init screen for UI
023.643  000.052: init default mappings
023.842  000.199: init default autocommands
030.110  004.590  004.590: sourcing /Users/robin/.local/share/nvim/site/autoload/plug.vim
044.500  000.029  000.029: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/elixir.vim
044.570  000.015  000.015: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/fennel.vim
044.636  000.019  000.019: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/fish.vim
044.696  000.015  000.015: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/fusion.vim
044.759  000.019  000.019: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/gdresource.vim
044.833  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/gdscript.vim
044.892  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/glimmer.vim
044.965  000.029  000.029: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/glsl.vim
045.020  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/gomod.vim
045.083  000.021  000.021: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/graphql.vim
045.145  000.019  000.019: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/hcl.vim
045.199  000.014  000.014: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/heex.vim
045.252  000.014  000.014: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/hjson.vim
045.307  000.016  000.016: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/json5.vim
045.371  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/julia.vim
045.431  000.019  000.019: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/ledger.vim
045.486  000.014  000.014: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/nix.vim
045.540  000.015  000.015: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/prisma.vim
045.593  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/pug.vim
045.655  000.019  000.019: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/ql.vim
045.716  000.021  000.021: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/query.vim
045.768  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/surface.vim
045.822  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/teal.vim
045.876  000.015  000.015: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/tlaplus.vim
045.930  000.015  000.015: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/yang.vim
045.984  000.015  000.015: sourcing /Users/robin/.vim/plugged/nvim-treesitter/ftdetect/zig.vim
046.394  008.856  008.423: sourcing /usr/local/share/nvim/runtime/filetype.vim
046.619  000.040  000.040: sourcing /usr/local/share/nvim/runtime/ftplugin.vim
046.836  000.034  000.034: sourcing /usr/local/share/nvim/runtime/indent.vim
047.208  000.086  000.086: sourcing /usr/local/share/nvim/runtime/syntax/synload.vim
047.298  000.362  000.276: sourcing /usr/local/share/nvim/runtime/syntax/syntax.vim
055.475  007.887  007.887: sourcing /Users/robin/.vim/plugged/gruvbox/colors/gruvbox.vim
128.540  104.452  082.682: sourcing /Users/robin/.config/nvim/init.vim
128.755  000.460: sourcing vimrc file(s)
129.192  000.149  000.149: sourcing /Users/robin/.vim/plugged/vim-commentary/plugin/commentary.vim
129.889  000.415  000.415: sourcing /Users/robin/.vim/plugged/vimwiki/autoload/vimwiki/vars.vim
131.001  000.101  000.101: sourcing /Users/robin/.vim/plugged/vimwiki/autoload/vimwiki/u.vim
133.204  003.906  003.390: sourcing /Users/robin/.vim/plugged/vimwiki/plugin/vimwiki.vim
133.377  000.015  000.015: sourcing /Users/robin/.vim/plugged/neoformat/plugin/neoformat.vim
133.512  000.023  000.023: sourcing /Users/robin/.vim/plugged/plenary.nvim/plugin/plenary.vim
133.742  000.124  000.124: sourcing /Users/robin/.vim/plugged/telescope.nvim/plugin/telescope.vim
136.402  002.530  002.530: sourcing /Users/robin/.vim/plugged/nvim-treesitter/plugin/nvim-treesitter.vim
136.855  000.125  000.125: sourcing /Users/robin/.vim/plugged/markdown-preview.nvim/plugin/mkdp.vim
138.559  001.515  001.515: sourcing /Users/robin/.vim/plugged/emmet-vim/plugin/emmet.vim
138.704  000.032  000.032: sourcing /Users/robin/.vim/plugged/vim-gist/plugin/gist.vim
138.888  000.056  000.056: sourcing /Users/robin/.vim/plugged/nvim-lspconfig/plugin/lspconfig.vim
139.161  000.028  000.028: sourcing /Users/robin/.vim/plugged/nvim-web-devicons/plugin/nvim-web-devicons.vim
139.947  000.135  000.135: sourcing /usr/local/share/nvim/runtime/plugin/gzip.vim
139.994  000.011  000.011: sourcing /usr/local/share/nvim/runtime/plugin/health.vim
140.066  000.037  000.037: sourcing /usr/local/share/nvim/runtime/plugin/man.vim
141.048  000.156  000.156: sourcing /usr/local/share/nvim/runtime/pack/dist/opt/matchit/plugin/matchit.vim
141.179  001.078  000.922: sourcing /usr/local/share/nvim/runtime/plugin/matchit.vim
141.358  000.143  000.143: sourcing /usr/local/share/nvim/runtime/plugin/matchparen.vim
141.413  000.019  000.019: sourcing /usr/local/share/nvim/runtime/plugin/netrwPlugin.vim
142.333  000.153  000.153: sourcing /usr/local/share/nvim/runtime/autoload/remote/host.vim
142.651  000.141  000.141: sourcing /usr/local/share/nvim/runtime/autoload/remote/define.vim
142.720  001.117  000.823: sourcing /Users/robin/.local/share/nvim/rplugin.vim
142.724  001.276  000.158: sourcing /usr/local/share/nvim/runtime/plugin/rplugin.vim
142.826  000.066  000.066: sourcing /usr/local/share/nvim/runtime/plugin/shada.vim
142.885  000.020  000.020: sourcing /usr/local/share/nvim/runtime/plugin/spellfile.vim
143.023  000.095  000.095: sourcing /usr/local/share/nvim/runtime/plugin/tarPlugin.vim
143.149  000.084  000.084: sourcing /usr/local/share/nvim/runtime/plugin/tohtml.vim
143.202  000.017  000.017: sourcing /usr/local/share/nvim/runtime/plugin/tutor.vim
143.385  000.144  000.144: sourcing /usr/local/share/nvim/runtime/plugin/zipPlugin.vim
144.432  000.381  000.381: sourcing /Users/robin/.vim/plugged/nvim-tree.lua/plugin/nvim-tree-startup.lua
146.515  001.703  001.703: sourcing /Users/robin/.vim/plugged/nvim-cmp/plugin/cmp.lua
147.080  004.617: loading rtp plugins
147.205  000.125: loading packages
147.812  000.235  000.235: sourcing /Users/robin/.vim/plugged/cmp-nvim-lsp/after/plugin/cmp_nvim_lsp.lua
148.296  000.377  000.377: sourcing /Users/robin/.vim/plugged/cmp-buffer/after/plugin/cmp_buffer.lua
148.961  000.278  000.278: sourcing /Users/robin/.vim/plugged/cmp-path/after/plugin/cmp_path.lua
149.182  001.087: loading after plugins
149.193  000.011: inits 3
153.349  004.156: reading ShaDa
155.506  000.396  000.396: sourcing /usr/local/share/nvim/runtime/autoload/dist/ft.vim
156.320  000.013  000.013: sourcing /Users/robin/.vim/plugged/nvim-treesitter/after/ftplugin/elixir.vim
156.444  000.012  000.012: sourcing /Users/robin/.config/nvim/after/ftplugin/elixir.vim
215.181  000.019  000.019: sourcing /Users/robin/.vim/plugged/nvim-treesitter/after/ftplugin/elixir.vim
215.331  000.015  000.015: sourcing /Users/robin/.config/nvim/after/ftplugin/elixir.vim
219.165  065.361: opening buffers
219.200  000.035: BufEnter autocommands
219.363  000.163: editing files in windows
219.384  000.020: VimEnter autocommands
219.386  000.002: UIEnter autocommands
219.796  000.246  000.246: sourcing /usr/local/share/nvim/runtime/autoload/provider/clipboard.vim
219.804  000.172: before starting main loop
1049.294  829.490: first screen update
1049.300  000.007: --- NVIM STARTED ---
```

**Comment by treeman** [comment]
I just noticed that `heex` files have horrendous performance, even small files are noticeably laggy.

**Comment by theHamsta** [comment]
@treeman If you're affected by the same bug, it's not about the files you're loading but certain queries that take eternities. Did you try to the bugfix by maxbrunsfeld?

**Comment by treeman** [comment]
@theHamsta I'm not sure which bugfix you're referring to, I tried to checkout neovim/neovim@64abd7b and it did improve things a lot.

Something I noticed is that starting up with a file `nvim app.html.heex` is fast, but looking at the file via telescope (finding files with preview) is still slow. It's not nearly as bad with the old commit, but even then it's noticeable. I only notice this for `.heex` files though, other elixir files are very fast.

This is an example of a slow file:

```
<main role="main" class="container">
  <p class="alert alert-info" role="alert"><%= get_flash(@conn, :info) %></p>
  <p class="alert alert-danger" role="alert"><%= get_flash(@conn, :error) %></p>
  <%= @inner_content %>
</main>
```

**Comment by jeroenbourgois** [comment]
@treeman you might consider disabling syntax highlighting in the preview window? That improved performance for me a lot.

**Comment by theHamsta** [comment]
@jeroenbourgois this should fix this issue for now in a way that you only have the parsing delay once: https://github.com/neovim/neovim/pull/16945

**Comment by jeroenbourgois** [comment]
Tnx!

Jeroen

**Comment by mfilej** [comment]
> @jeroenbourgois this should fix this issue for now in a way that you only have the parsing delay once: [neovim/neovim#16945](https://github.com/neovim/neovim/pull/16945)

Can confirm. The delay is gone afterwards, but is still present on first load. Is there an issue we can track?

**Comment by kiuKisas** [comment]
Weird, it seems that it's not properly fix (slow at start time).. :(

**Comment by mfilej** [comment]
Personally I am unable to accurately report any performance issues because I switched to a new computer during the time when this issue was open. So while I do notice that opening the buffer for the first time is slower than opening it subsequently, I can't really say whether it is slower than it used to be before this issue first appeared. Will open a new issue if/when I have a good comparison.

Thanks to the maintainers for the hard work!
